### DISCUSSION TITLE
Type narrowing after use of all() built-in

### DISCUSSION THREAD

**Discussion by FinemechanicPub** [microsoft/pyright#8665] [Ideas]
I think it would be great if Pyright could detect that a sequence can't contain `None` after a check with `all()`:

```python
from typing import Sequence


def func(nums: Sequence[int | None]):
    if not all(nums):
        raise ValueError
    total = 0
    for num in nums:
        total += num
    return total
```

**Comment by erictraut** [comment]
This isn't currently a supported type guard pattern. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

I've seen the `all(x)` form used in real-world code, but the type of `x` can rarely be narrowed because it's most often an invariant container type like `list` or `set`. Your code example uses a `Sequence`, which is covariant. That means it could work in this case; it would be safe to narrow `Sequence[int | None]` to `Sequence[int]`. But I'm not convinced that adding the special support for this pattern is worth it because it wouldn't work very often in real code.

I recommend using a user-defined type guard (with the `TypeIs` special form) to handle your use case.

```python
from typing import Sequence, TypeIs

def no_nones_or_zeros[T](val: Sequence[T | None]) -> TypeIs[Sequence[T]]:
    return all(val)

def func(nums: Sequence[int | None]):
    if not no_nones_or_zeros(nums):
        raise ValueError
    total = 0
    for num in nums:
        total += num
    return total
```

**Reply by FinemechanicPub** [reply]
Thank you for advice.
